### PR TITLE
Modularize realtime inference

### PR DESCRIPTION
**Pull request by zhiqiangdon**
This PR is to modularize some logic shared by the realtime inference and dataloader. This modularization can also benefit the ONNIX export by @FANGAreNotGnu.

1. Add util `apply_df_preprocessor` shared by the realtime inference and dataloader.
2. Add util `apply_data_processor` shared by the realtime inference and dataloader.
3. Add util `infer_batch` for realtime inference.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by bryanyzhu — APPROVED** [review]
Great!

File: multimodal/src/autogluon/multimodal/utils.py [hunk 1]
```diff
@@ -2781,3 +2782,40 @@ def move_to_device(obj: Union[torch.Tensor, nn.Module, Dict, List], device: torc
...
+    batch_size = len(batch[next(iter(batch))])
+    if 1 < torch.cuda.device_count() <= batch_size:
+        model = nn.DataParallel(model)
+    model.to(device).eval()
+    batch = move_to_device(batch, device=device)
+    with torch.autocast(device_type=device_type, dtype=precision):
+        with torch.no_grad():
+            output = model(batch)[model.prefix]
+
+    if isinstance(loss_func, nn.BCEWithLogitsLoss):
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/utils.py:2818]: Should we put it under `torch.no_grad()?`

File: multimodal/src/autogluon/multimodal/utils.py [hunk 2]
```diff
@@ -2781,3 +2782,40 @@ def move_to_device(obj: Union[torch.Tensor, nn.Module, Dict, List], device: torc
...
+    precision
+        The desired precision used in inference.
+    loss_func
+        The loss function used in training the model.
+
+    Returns
+    -------
+    Model output.
+    """
+    precision = infer_precision(num_gpus=1, precision=precision, as_torch=True)
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/utils.py:2806]: When reading the code, I think we should set `num_gpus` to `toch.cuda.device_count()`.

File: multimodal/src/autogluon/multimodal/utils.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: Done.

File: multimodal/src/autogluon/multimodal/utils.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: Good suggestion! Updated.

File: multimodal/src/autogluon/multimodal/utils.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: BTW, I am thinking about a form way to handle this kind of post-processing like `nn.BCEWithLogitsLoss` . The current way is kind of a hack and have several repeated code snippets, which is not easy to maintain.

**Review by suzhoum — APPROVED** [review]
LGTM! Thanks for answering my questions!

**Pull request merged**